contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,350 @@
+The control plane upgrade is atomic; either all or nocontrol plane components get upgraded. If a post-upgrade step fails though (like upgrading Bootstrap Tokens or applying a new kube-dns), kubeadm upgrade just returns a non-zero code instead of reverting the successful control plane upgrade. We might change the latter behavior in the future.
+
+In the first iteration, upgrading the kubelets and/or etcd in the cluster is a manual task (see the sections below).
+We will address these limitations in future versions before graduating this functionality to GA.
+
+## Command structure
+
+ - `kubeadm upgrade`: Placeholder command for other subcommands.
+   - Relevant, generic flags:
+     - `--config`: A path to a configuration file to use as the desired state for the cluster component configuration. If omitted, configuration is read from a ConfigMap in the cluster.
+     - `--kubeconfig`: The path to a configuration file to use when talking to the cluster. Default: `/etc/kubernetes/admin.conf`.
+     - `--allow-experimental-upgrades`: Allow the user to upgrade to alpha, beta and release candidate versions.
+     - `--allow-release-candidate-upgrades`: Allow the user to upgrade to release candidate versions.
+     - `--print-config`: Print out the configuration read from the file or ConfigMap in the cluster.
+     - `--skip-preflight-checks`: Skip the preflight checks that are run before execution.
+ - `kubeadm upgrade plan`: Serves as a way to show users:
+   - If the cluster is in an upgradeable state.
+   - What versions the user can upgrade the cluster to.
+ - `kubeadm upgrade apply [version]`:
+   - Is the command that actually performs the upgrade (of the control plane).
+   - Will interactively ask the user for a `[Y/n]` confirmation if the upgrade really should be performed.
+   - Relevant flags:
+     - `--dry-run`: Only log the changes that would be made to the cluster without changing any state.
+     - `--force`, `-f`: Ignore any skippable version policy errors and the interactive yes/no upgrade confirmation.
+     - `--image-pull-timeout`: The time to wait for the new control plane images to pull before timing out. Default: 15 minutes.
+     - `--yes`, `-y`: Run the command non-interactively, don't ask for an upgrade confirmation.
+
+## Example user experience
+
+UX when upgrading a patch version:
+
+```console
+$ kubeadm upgrade plan
+[kubeadm upgrade plan output]
+```
+
+To upgrade the kubelets in the cluster, please upgrade them from the source you installed them from (most often using your package manager).
+
+```console
+$ kubeadm upgrade apply --version v1.8.3
+[kubeadm upgrade apply output]
+```
+
+## Upgrade planning
+
+In order for the user to know whether the cluster may be upgraded or not; the `kubeadm upgrade plan` command was written.
+The command:
+
+1. Checks if the cluster is healthy
+  - Does the API server `/healthz` endpoint return `ok`?
```

Comment:
we don't check the other master components as well? i know that the componentstatuses endpoint is a terrible hack but it does tell you whether the scheduler and controller manager (and etcd if it's at localhost) are also healthy.

Reply:
We don't right now, as I don't know how reliable that actually is.

Reply:
componentstatuses is pretty reliable the way we have the master configured in gke. i don't know if that translates to it being reliable in general though.